### PR TITLE
Class new line between docstrings / vars / methods

### PR DESCRIPTION
**Pull request by Sterbic**
Input:
```
class ClassWithInit:
    def __init__(self):
        pass
```

Current output:
```
class ClassWithInit:
+
    def __init__(self):
        pass
```

Desired output:
```
class ClassWithInit:
    def __init__(self):
        pass
```

Input:
```
class ClassWithDocstring:
    """docstring"""
    var = 10
    def __init__(self):
        pass
```

Current output:
```
class ClassWithDocstring:
    """docstring"""
    var = 10

    def __init__(self):
        pass
```

Desired output:
```
class ClassWithDocstring:
    """docstring"""

    var = 10

    def __init__(self):
        pass
```

### CODE REVIEW

File: tests/class_methods_new_line.py [hunk 1]
```diff
@@ -0,0 +1,42 @@
...
+
+# output
+
+
+class ClassSimplest:
+    pass
+
+
+class ClassWithInit:
+
```
Comment (Sterbic) [inline tests/class_methods_new_line.py]: This new line needs to be removed

File: tests/class_methods_new_line.py [hunk 2]
```diff
@@ -0,0 +1,42 @@
...
+class ClassWithInitAndVars:
+    cls_var = 100
+
+    def __init__(self):
+        pass
+
+
+class ClassWithInitAndVarsAndDocstring:
+    """Test class"""
+    cls_var = 100
```
Comment (Sterbic) [inline tests/class_methods_new_line.py]: @ambv, I'd usually put an empty line between docstrings and the rest of the code below. What's your view on this?

Comment (ambv) [reply tests/class_methods_new_line.py]: Black currently does not enforce that. What does PEP 257 say?

Comment (Sterbic) [reply tests/class_methods_new_line.py]: > There's no blank line either before or after the docstring.

> Insert a blank line after all docstrings (one-line or multi-line) that document a class -- generally speaking, the class's methods are separated from each other by a single blank line, and the docstring needs to be offset from the first method by a blank line.

(src: https://www.python.org/dev/peps/pep-0257/)

File: black.py [hunk 3]
```diff
@@ -909,6 +909,13 @@ def is_class_paren_empty(self) -> bool:
             and self.leaves[3].value == ")"
         )
 
+    @property
+    def is_docstring(self) -> bool:
+        """Is the line a triple quoted docstring?"""
```
[the remark this answers is not in the document]
Comment (Sterbic) [reply black.py]: 1. With this you mean that I should check the ending quotes as well?
2. I don't think `is_multiline_string` is applicable here. The docstring can be in a single line but we would still want an empty line after it and `is_multiline_string` returns `False` for `"""test"""`.
3. Will fix.

**Comment by Sterbic** [comment]
@ambv, anything else I should take care of here? PEP 257 is now enforced and class def and first method are not separated by a new line any more.

**Comment by renstrom** [comment]
Sorry for the too-late comment. But I think this is wrong, and the old behaviour was correct.

>Desired output:
>
>```
>class ClassWithInit:
>    def __init__(self):
>        pass
>```

For one, having a blank line is more consistent with the behaviour you get if you have a class docstring. And second, it's consistent with the PEP8 rule the methods should be surrounded by a blank line.

**Comment by renstrom** [comment]
PEP 8 also says this:

>Method definitions inside a class are surrounded by a single blank line.

It doesn't go into much detail, but it doesn't mention any exception like the `class` definition. Therefore I think it should be treated as there should _always_ be blank line before and after the method.

**Comment by Sterbic** [comment]
I believe this should be consistent regardless of what the first line after class declaration is, be it a method, variable or docstring. And you wouldn't put an empty line between class and docstring, as per PEP 257.

If you check the wording in PEP 257 it actually clarifies that the one empty line rule is meant to be applied between 2 methods only:
> ... generally speaking, the class's methods are separated from **each other** by a single blank line
(src: https://www.python.org/dev/peps/pep-0257/)

Various PEPs use this style in examples, including PEP 8, PEP 484 and PEP 525.

**Comment by ambv** [comment]
@renstrom, if you decide to deconstruct particular wording in PEP 8 and PEP 257, you might be technically right but I don't think that would be right in the spirit of those documents, they aren't written in a particularly strict style (pun not intended). The authors of those PEPs write code that looks like what Luka implemented in this PR. Previous behavior was very unpopular and a big criticism of Black for "putting unnecessary empty lines" between the first method and the class signature.

**Pull request merged**